pyscriptjs/src/plugins/pyterminal.ts

```diff
@@ -89,38 +130,117 @@ function make_PyTerminal(app: PyScriptApp) {
             this.outElem.className = 'py-terminal';
             this.appendChild(this.outElem);
 
-            if (this.isAuto()) {
-                this.classList.add('py-terminal-hidden');
-                this.autoShowOnNextLine = true;
-            } else {
-                this.autoShowOnNextLine = false;
-            }
+            this.setupPosition(app);
+        }
 
+        // implementation of the Stdio interface
+        stdout_writeline(msg: string) {
+            this.outElem.innerText += msg + '\n';
             if (this.isDocked()) {
-                this.classList.add('py-terminal-docked');
+                this.scrollTop = this.scrollHeight;
+            }
+            if (this.autoShowOnNextLine) {
+                this.classList.remove('py-terminal-hidden');
+                this.autoShowOnNextLine = false;
             }
+        }
 
-            logger.info('Registering stdio listener');
-            app.registerStdioListener(this);
+        stderr_writeline(msg: string) {
+            this.stdout_writeline(msg);
         }
+        // end of the Stdio interface
+    }
 
-        isAuto() {
-            return this.hasAttribute('auto');
+    return PyTerminalPre;
+}
+
+declare const Terminal: typeof TerminalType;
+
+function make_PyTerminal_xterm(app: PyScriptApp) {
+    /** The <py-terminal> custom element, which automatically register a stdio
+     *  listener to capture and display stdout/stderr
+     */
+    class PyTerminalXterm extends PyTerminalBaseClass {
+        outElem: HTMLDivElement;
+        moduleResolved: boolean;
+        xtermReadyPromise: Promise<TerminalType>;
+        xterm: TerminalType;
+        cachedStdOut: Array<string>;
+        cachedStdErr: Array<string>;
+
+        constructor() {
+            super();
+            this.moduleResolved = false;
+            this.cachedStdOut = [];
+            this.cachedStdErr = [];
+
+            //Required to make xterm appear properly
+            this.style.width = '100%';
+            this.style.height = '100%';
         }
 
-        isDocked() {
-            return this.hasAttribute('docked');
+        async connectedCallback() {
+            this.outElem = document.createElement('div');
+            //this.outElem.className = 'py-terminal';
+            this.appendChild(this.outElem);
+
+            this.setupPosition(app);
+
+            this.xtermReadyPromise = this.setupXterm();
```

Comment:
if this is how users should await the terminal I feel like `setupXterm` shouldn't be a public method, rather an external private funciton/utility to `.call(this)` or at least a `_setupXterm` prefixed method because it's otherwise exposed as API nobody should use when `connectedCallback` has not happened ... can we put at least a `@private` hint to the method or make it more obvious it's an internal implementation detail?

Reply:
Ah, `moduleResolved` is there for the purposes of having a synchronous way of checking whether the xterm.js module is initialized and ready to be written to. If `stdout_writeline` or `stderr_writeline` are called (i.e. Python is up and running), but the xterm.js module is not ready yet, we want to cache those writes, and play them back when the xterm.js is ready. If we just awaited the whenReady promise, `stdout_writeline` and `stderr_writeline` would have be async... and I'm not sure what the ramifications of that would be.